diff -Naur og.orig/og_ui/og_ui.module og/og_ui/og_ui.module
--- og.orig/og_ui/og_ui.module	2014-10-27 16:43:19.000000000 +0100
+++ og/og_ui/og_ui.module	2014-10-27 16:45:02.000000000 +0100
@@ -694,7 +694,9 @@
       }
 
       if (!empty($links['title'])) {
-        $links += array('options' => array('attributes' => array('title' => $links['title'], 'class' => array($links['class']))));
+        // Patch upstream, we need array_merge_recursive here.
+        $links_options = array('options' => array('attributes' => array('title' => $links['title'], 'class' => array($links['class']))));
+        $links = array_merge_recursive($links, $links_options);
         $element[0] = array(
           '#type' => 'link',
           '#title' => $links['title'],
