--- a/includes/install.inc	2016-07-14 09:59:45.000000000 +0200
+++ b/includes/install.inc	2016-07-14 10:04:56.000000000 +0200
@@ -1180,9 +1180,6 @@

   $info = install_profile_info($profile);

-  // Add the profile as a dependency itself, in order to see if it has requirements as well.
-  $info['dependencies'][] = $profile;
-
   // Collect requirement testing results.
   $requirements = array();
   foreach ($info['dependencies'] as $module) {
