diff --git a/drupal-org.make.orig b/drupal-org.make
index e4e0fa5..ad9fffd 100644
--- a/drupal-org.make.orig
+++ b/drupal-org.make
@@ -58,7 +58,7 @@ projects[apachesolr_user][download]revision] = "a86c5aebfceaf4a3fc53544762a36ca1

 ; Check the user object before trying to display a result.
 ; https://drupal.org/node/2077281#comment-7807937
-projects[apachesolr_user][patch][] = "https://drupal.org/files/2077281-apache-solr-user-check-3.patch"
+;projects[apachesolr_user][patch][] = "https://drupal.org/files/2077281-apache-solr-user-check-3.patch"

 projects[breakpoints][type] = "module"
 projects[breakpoints][subdir] = "contrib"
