diff -Naur a/drupal-org.make b/drupal-org.make
--- a/drupal-org.make	2017-02-01 17:24:11.000000000 +0100
+++ b/drupal-org.make	2017-02-01 17:31:33.000000000 +0100
@@ -426,7 +426,7 @@

 projects[title][type] = "module"
 projects[title][subdir] = "contrib"
-projects[title][version] = "1.0-alpha8"
+projects[title][version] = "1.0-alpha9"

 projects[token][type] = "module"
 projects[token][subdir] = "contrib"
