diff -Naur commons.orig/drupal-org.make commons/drupal-org.make
--- commons.orig/drupal-org.make	2017-02-27 15:04:30.000000000 +0100
+++ commons/drupal-org.make	2017-02-27 15:05:42.000000000 +0100
@@ -442,7 +442,7 @@
 
 projects[views][type] = "module"
 projects[views][subdir] = "contrib"
-projects[views][version] = "3.14"
+projects[views][version] = "3.15"
 
 ; Update Views Content access filter per core performance improvements.
 ; https://drupal.org/comment/8516039#comment-8516039
